src/debug_sdfs_ed64.c

```diff
@@ -0,0 +1,598 @@
+		if (!num_retries--) {
+			debugf("SD card did not respond\n");
+			return RES_ERROR;
+			break;
+		}
+
+		// Query with HCS and 3.2-3.4V
+		if (everdrive_sd_send_app_command(ED64_SD_CMD41, 0, 0x40300000)) {
+			// Check ready bit on OCR
+			if ((everdrive_sd_resp_buff[1] & 0x80) != 0) break;
+		}
+
+		wait_ms(ED64_SD_ACMD41_WAIT_MS);
+	};
+
+	// Check CCS and set HC mode
+	everdrive_sd_active_mode |= everdrive_sd_resp_buff[1] & 0x40;
+
+	if (!everdrive_sd_execute_command(ED64_SD_CMD2, 0)) return RES_ERROR;
+
+	if (!everdrive_sd_execute_command(ED64_SD_CMD3, 0)) return RES_ERROR;
+
+	sd_rca = 	(everdrive_sd_resp_buff[1] << 24) |
+				(everdrive_sd_resp_buff[2] << 16) |
+				(everdrive_sd_resp_buff[3] << 8) |
+				everdrive_sd_resp_buff[4];
+
+	if (!everdrive_sd_execute_command(ED64_SD_CMD7, sd_rca)) return RES_ERROR;
+
+	// Set bus width to 4
+	if (!everdrive_sd_send_app_command(ED64_SD_CMD6, sd_rca, 0x2)) {
+		debugf("ACMD6 err\n");
+		return RES_ERROR;
+	};
+
+	// Set hi speed
+	everdrive_sd_config |= ED64_SD_CFG_SPEED;
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_STATUS, everdrive_sd_config);
+
+	return RES_OK;
+}
+
+static DRESULT fat_disk_read_everdrive(BYTE* buff, LBA_t sector, UINT count)
+{
+	_Static_assert(FF_MIN_SS == 512, "this function assumes sector size == 512");
+	_Static_assert(FF_MAX_SS == 512, "this function assumes sector size == 512");
+
+	uint8_t crc[8];
+	DRESULT ret_val = RES_OK;
+
+	// Overclock the PI
+	uint32_t old_pw = PI_regs->dom1_pulse_width;
+	io_write((uint32_t)&PI_regs->dom1_pulse_width, 0x09);
+
+	if (!everdrive_sd_change_mode(ED64_SD_MODE_BLOCK_READ, sector)) {
+		ret_val = RES_ERROR; goto cleanup;
+	};
+
+	for (int i=0;i<count;i++)
+	{
+		uint16_t timeout = -1;
+		uint8_t result;
+		// Each 1 bit everdrive_sd_read_data shifts 4 bit of data from the four
+		// data lanes. To find the start marker, wait for all lanes to go low to
+		// start the transfer.
+		set_everdrive_sd_bitlen(1);
+		do {
+			if (!timeout--) {
+				debugf("Data token timeout\n");
+				ret_val = RES_ERROR; goto cleanup;
+			}
+			result = everdrive_sd_read_data();
+		} while (result != 0xf0);
+
+		// It is also possible to read the data one byte at a time but this is
+		// inefficient - left here for documentation only.
+		// set_everdrive_sd_bitlen(2); // Read a byte from wide bus each time
+		// for(int j = 0; j < 512; j++) {
+		// 	buff[j] = everdrive_sd_read_data();
+		// }
+
+		set_everdrive_sd_bitlen(4);
+
+		data_cache_hit_writeback_invalidate(buff, 512);
+		dma_read(buff,  ED64_BASE_ADDRESS + ED64_SD_IO_BUFFER, 512);
+
+		// TODO: actually check the CRC?
+		data_cache_hit_writeback_invalidate(crc, 8);
+		dma_read(crc, ED64_BASE_ADDRESS + ED64_SD_IO_BUFFER, 8);
+		buff += 512;
+	}
+
+	everdrive_sd_address = sector + count;
+
+cleanup:
+	io_write((uint32_t)&PI_regs->dom1_pulse_width, old_pw);
+	if (ret_val == RES_OK) return RES_OK;
+
+	// Do error cleanup
+	everdrive_sd_change_mode(ED64_SD_MODE_NONE, 0);
+
+	return ret_val;
```

Comment:
In this cleanup, you don't change `everdrive_sd_address`. It looks to me that if you call this function to read 4 blocks, and after 2 blocks the reading goes into timeout, `everdrive_sd_address` will still point to the initial block, while in reality two blocks were read and the current address is probably the third block. Does this make any sense?

I'm worried in general of these global caches. I mean they're fine to minimize redundant hardware accesses but they must be very carefully updated on every code path.

Reply:
Yes, that makes sense but `everdrive_sd_change_mode` already sets `everdrive_sd_address` to zero and changes the mode into `ED64_SD_MODE_NONE` which always require a sector restart anyways. `everdrive_sd_change_mode` will also stop the transmission in that case. Here I assume if something breaks, we have lost the current sector and start over.

In general I agree but especially for the multi block transmission, we have to keep the state as it is aimed at reducing successive SD commands. o/w we would use single block reads and update the sector before each read.